runtime/intern.c

```diff
@@ -907,11 +904,11 @@ CAMLexport value caml_input_val_from_bytes(value str, intnat ofs)
   caml_parse_header(s, "input_val_from_string", &h);
   if (ofs + h.header_len + h.data_len > caml_string_length(str))
     caml_failwith("input_val_from_string: bad length");
-  /* Decompress if needed */
-  s->intern_src = &Byte_u(str, ofs + h.header_len); /* If a GC occurred */
-  intern_decompress_input(s, "input_val_from_string", &h);
   /* Allocate result */
   intern_alloc_storage(s, h.whsize, h.num_objects);
+  s->intern_src = &Byte_u(str, ofs + h.header_len); /* If a GC occurred */
+  /* Decompress if needed */
+  intern_decompress_input(s, "input_val_from_string", &h);
```

Comment:
Yes, this is obviously more correct :-)  and there's no regrets w.r.t. memory usage, since `str` is in the OCaml heap (it's not malloc-ed) and live throughout the function, so `intern_decompress_input` has no chance to free it.